Make it possible to turn caching off per table and stop caching CDC Log

haaawk opened the pull request:
We inherited from Origin a `caching` table parameter. It's a map of named caching parameters. Before this PR two caching parameters were expected: `keys` and `rows_per_partition`. So far we have been ignoring them. This PR adds a new caching parameter called `enabled` which can be set to `true` or `false` and controls the usage of the cache for the table. By default, it's set to `true` which reflects Scylla behavior before this PR.

This new capability is used to disable caching for CDC Log table. It is desirable because CDC Log entries are not expected to be read often. They also put much more pressure on memory than entries in Base Table. This is caused by the fact that some writes to Base Table can override previous writes. Every write to CDC Log is unique and does not invalidate any previous entry.

Fixes #6098
Fixes #6146

Tests: unit(dev, release), manual

denesb commented on caching_options.hh:
```diff
@@ -59,8 +62,17 @@ class caching_options {
+        std::map<sstring, sstring> res = {{ "keys", _key_cache },
+                { "rows_per_partition", _row_cache }};
+        if (!_enabled) {
+            res.insert({"enabled", "false"});
```
Why is this conditional?

haaawk replied:
Because missing parameter means enabled == true. That's just to keep it more backward compatible.

denesb replied:
I think we should add it unconditionally, otherwise all code wishing to use this map has to learn about what a missing `enabled` means.

haaawk replied:
The code will have to do it anyway because the parameter is missing on older nodes and then it means enabled == true.

denesb replied:
Still, why complicate this method for no gain? We don't need this to be backward compatible as older nodes will not see this option anyway, will they?

haaawk replied:
Well they shouldn't see it because of the feature but I think it's safer for rollback or stuff like this. This is not big complication and actually gives us more safety. For example if no table has disabled cache, it is like the feature does not exist at all. zero-overhead feature ;)

denesb replied:
Ok.

avikivity replied:
Agree it's safer. It also allows a DESCRIBE TABLE on a new cluster to be replayed on an old cluster with fewer surprises.

avikivity replied:
It's also better to preserve the schema digest.

avikivity replied:
But, something is missing here. If the user specifed enabled: true explicitly, then when we call to_map() we get it back without the option.

You can fix by making _enabled an optional<boo> (and return value_or(true) from enabled()).

avikivity replied:
And this can lead to schema digest mismatch. Please add a test that fails on the current patch and succeeds on the fixed patch.

haaawk replied:
Could you explain more how this can create a schema digest mismatch @avikivity? I don't see it. No matter whether enabled was set explicitly or wasn't set at all, exchanged schema will have no 'enabled' element in the map. Or am I missing something?

haaawk replied:
> But, something is missing here. If the user specifed enabled: true explicitly, then when we call to_map() we get it back without the option.
> 
> You can fix by making _enabled an optional (and return value_or(true) from enabled()).

Setting 'enabled' to true and not setting means the same. Is it really a problem?

avikivity replied:
I think you are correct. I thought the map could get into schema_tables directly, but it always goes through to_map(), so enabled: true will always be filtered out.

avikivity commented on table.cc:
```diff
@@ -723,8 +723,7 @@ table::update_cache(lw_shared_ptr<memtable> m, sstables::shared_sstable sst) {
     } else {
-        adder();
-        return m->clear_gently();
+        return _cache.invalidate(adder).then([m] { return m->clear_gently(); });
```
Why do we need the clear_gently?

haaawk replied:
I'm not sure. I just left what was there before. @tgrabiec probably knows best.

avikivity replied:
Ah, of course you need it, to remove stuff from the memtable. I didn't look at the code thoroughly and thought it was about the cache.

avikivity commented on database.cc:
```diff
@@ -906,7 +906,7 @@ keyspace::make_column_family_config(const schema& s, const database& db) const {
     cfg.enable_disk_writes = _config.enable_disk_writes;
     cfg.enable_commitlog = _config.enable_commitlog;
-    cfg.enable_cache = _config.enable_cache;
+    cfg.enable_cache = _config.enable_cache && s.caching_options().enabled();
```
What about ALTER TABLE? We need to be able to change this dynamically.

haaawk replied:
Alter table works as well. I checked.

haaawk replied:
Well - I checked that the property was set, I will double-check that new operations on the table start/stop using cache.

haaawk replied:
You're right. The alter table does not change the behavior dynamically. Only after a restart. I will fix that.

avikivity commented:
Looks good, but need to handle ALTER TABLE.

haaawk commented:
I added support for ALTER TABLE @avikivity and tested it both enable -> disable and disable -> enable.

denesb commented:
@haaawk what about `schema::describe()`?

haaawk commented:
> @haaawk what about `schema::describe()`?

It just works @denesb - WITH CACHING was there before. I just added new param inside the map.

denesb commented:
> > @haaawk what about `schema::describe()`?
> 
> It just works @denesb - WITH CACHING was there before. I just added new param inside the map.

Ok, LGTM then. We should also check how cqlsh et. al. handles the new param, but that is different repo anyway.

haaawk commented:
> > > @haaawk what about `schema::describe()`?
> > 
> > 
> > It just works @denesb - WITH CACHING was there before. I just added new param inside the map.
> 
> Ok, LGTM then. We should also check how cqlsh et. al. handles the new param, but that is different repo anyway.

I tried it and it just prints it alright. It is simple because this is a standard param map used everywhere (compaction, network strategy, ...)

denesb commented:
On Thu, 2020-04-30 at 08:19 -0700, Piotr Jastrzębski wrote:
> 
> 
> > > > @haaawk what about schema::describe()?
> > > 
> > > It just works @denesb - WITH CACHING was there before. I just
> > > added new param inside the map.
> > 
> > Ok, LGTM then. We should also check how cqlsh et. al. handles the
> > new param, but that is different repo anyway.
> 
> I tried it and it just prints it alright. It is simple because this
> is a standard param map used everywhere (compaction, network
> strategy, ...)

Awesome!



—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub, or unsubscribe.
[
{
"@context": "http://schema.org",
"@type": "EmailMessage",
"potentialAction": {
"@type": "ViewAction",
"target": "https://github.com/scylladb/scylla/pull/6312#issuecomment-621920932",
"url": "https://github.com/scylladb/scylla/pull/6312#issuecomment-621920932",
"name": "View Pull Request"
},
"description": "View this Pull Request on GitHub",
"publisher": {
"@type": "Organization",
"name": "GitHub",
"url": "https://github.com"
}
}
]

haaawk commented:
ping @avikivity @tgrabiec

avikivity commented on cql3/statements/cf_prop_defs.hh:
```diff
@@ -95,6 +95,7 @@ public:
     std::map<sstring, sstring> get_compaction_options() const;
     std::optional<std::map<sstring, sstring>> get_compression_options() const;
     const cdc::options* get_cdc_options(const schema::extensions_map&) const;
+    std::optional<caching_options> get_caching_options() const;
```
Maybe it's simpler not to return optional here, a map can be empty. Returning an optional means everyone has to check if it exists or not.

haaawk replied:
If I remember correctly, this function is used in both create table and alter table. In alter table, an empty map is different than not present at all. The empty map clears out the setting while the missing map leaves the setting unchanged.

avikivity replied:
Makes sense!

avikivity commented on table.cc:
```diff
@@ -2008,6 +2008,10 @@ void table::set_schema(schema_ptr s) {
+    _config.enable_cache = ks.get_config().enable_cache && _schema->caching_options().enabled();
+    if (!_config.enable_cache) {
+        _cache.invalidate([] {}).get();
+    }
```
Reads that are in-progress while this is happening can still populate the cache after invalidate() completes.

That's not a huge problem since the cache will still be coherent, and when we enable it again the data from those reads will be valid.

Maybe it's better to drop the invalidate() call. The data in the cache will age out naturally, and this allows ALTER TABLE to complete more quickly.

haaawk replied:
I see no problem in removing this. Will do.

avikivity commented:
Looks good. I think I found a minor problem in to_map().

haaawk commented:
> Looks good. I think I found a minor problem in to_map().

Is it really a problem that to_map always returns missing 'enabled' when it's set to true? It's safer regarding backward compatibility this way.

haaawk commented:
I removed cache invalidation in alter table. @avikivity @tgrabiec please have a look

haaawk commented on table.cc:
```diff
@@ -1951,7 +1951,7 @@ std::chrono::steady_clock::duration table::enable_sstable_write(int64_t new_gene
 }
 
-void table::set_schema(schema_ptr s) {
+void table::set_schema(const keyspace& ks, schema_ptr s) {
```
Sure. I will give it a try and let you know if it worked out well.

haaawk replied:
done.

haaawk commented:
I addressed @tgrabiec comment and added table::cache_enabled(). @avikivity @tgrabiec Do you have any more comments or are we good to merge?

tgrabiec commented:
@haaawk I don't see `cache_enabled()` in the branch.

haaawk commented:
> @haaawk I don't see `cache_enabled()` in the branch.

I pushed the wrong branch. Sorry about that @tgrabiec. I'm pushing the correct one right now.

haaawk commented:
@avikivity Tomek and Botond already approved. Would you do the honors?

avikivity commented:
Queued.

haaawk commented:
> Queued.

Thank you.

avikivity commented:
Dequeued per @bhalevy's request, fails next-gating dtests.

haaawk commented:
I had to add property_definitions::get which returns variant and allows checking whether the parameter has a map or a string value and use it to handle legacy representation of caching parameter. @avikivity please have a look.

haaawk commented:
@avikivity for your convenience here's the diff on the change since last version that was approved:
```
diff --git a/cql3/statements/property_definitions.hh b/cql3/statements/property_definitions.hh
index 5cfe01450..0b51d9de6 100644
--- a/cql3/statements/property_definitions.hh
+++ b/cql3/statements/property_definitions.hh
@@ -86,6 +86,8 @@ class property_definitions {
 public:
     bool has_property(const sstring& name) const;
 
+    std::optional<value_type> get(const sstring& name) const;
+
     sstring get_string(sstring key, sstring default_value) const;
 
     // Return a property value, typed as a Boolean
diff --git a/cql3/statements/cf_prop_defs.cc b/cql3/statements/cf_prop_defs.cc
index 5fffd399c..8c38f15e0 100644
--- a/cql3/statements/cf_prop_defs.cc
+++ b/cql3/statements/cf_prop_defs.cc
@@ -205,11 +205,18 @@ std::optional<utils::UUID> cf_prop_defs::get_id() const {
 }
 
 std::optional<caching_options> cf_prop_defs::get_caching_options() const {
-    auto caching_options_map = get_map(KW_CACHING);
-    if (!caching_options_map || caching_options_map->empty()) {
+    auto value = get(KW_CACHING);
+    if (!value) {
         return {};
     }
-    return caching_options::from_map(*caching_options_map);
+    return std::visit(make_visitor(
+        [] (const property_definitions::map_type& map) {
+            return map.empty() ? std::nullopt : std::optional<caching_options>(caching_options::from_map(map));
+        },
+        [] (const sstring& str) {
+            return std::optional<caching_options>(caching_options::from_sstring(str));
+        }
+    ), *value);
 }
 
 const cdc::options* cf_prop_defs::get_cdc_options(const schema::extensions_map& schema_exts) const {
diff --git a/cql3/statements/property_definitions.cc b/cql3/statements/property_definitions.cc
index 39f605fee..9cbdce86f 100644
--- a/cql3/statements/property_definitions.cc
+++ b/cql3/statements/property_definitions.cc
@@ -109,6 +109,13 @@ bool property_definitions::has_property(const sstring& name) const {
     return _properties.find(name) != _properties.end();
 }
 
+std::optional<property_definitions::value_type> property_definitions::get(const sstring& name) const {
+    if (auto it = _properties.find(name); it != _properties.end()) {
+        return it->second;
+    }
+    return std::nullopt;
+}
+
 sstring property_definitions::get_string(sstring key, sstring default_value) const {
     auto value = get_simple(key);
     if (value) {
```

haaawk commented:
Do you have any more comments @avikivity?

avikivity commented:
Looks good, merging.
